lints/etsi/lint_qcstatem_psd2_policy_mandatory.go

```diff
@@ -0,0 +1,80 @@
+	isPresent := util.HasCertAnyEtsiQcStatement(c)
+	return isPresent
+}
+
+func policyAndTypeAreConsistent(c *x509.Certificate, policy asn1.ObjectIdentifier, policyStr string, qcType asn1.ObjectIdentifier, typeStr string, mustHaveQcStmtFour bool) string {
+
+	if util.HasCertPolicy(c, policy) && !util.HasCertEtsiQcType(c, qcType) {
+		return "EU Qualified Certificate has policy " + policyStr + " but doesn't have the corresponding QcType '" + typeStr + "' (ETSI EN 319 412-4: \"Policy identifiers included in the certificate policies extension of EU Qualified Certificates shall be consistent with the EU Qualified Certificate Statements\")"
+	}
+	_, isQscdStmtPresent := util.IsQcStatemPresent(c, &util.IdEtsiQcsQcSSCD)
+	if util.HasCertPolicy(c, policy) && mustHaveQcStmtFour && !isQscdStmtPresent {
+		return "EU Qualified Certificate has policy indicating private key storage in QSCD, but the corresponding QC Statement 4 is missing (ETSI EN 319 412-4: \"Policy identifiers included in the certificate policies extension of EU Qualified Certificates shall be consistent with the EU Qualified Certificate Statements\")."
+	}
+	if util.HasCertPolicy(c, policy) && !mustHaveQcStmtFour && isQscdStmtPresent {
+		return "EU Qualified Certificate does not feature the policy indicating private key storage in QSCD, but the corresponding QC Statement 4, indicating this, is present (ETSI EN 319 412-4: \"Policy identifiers included in the certificate policies extension of EU Qualified Certificates shall be consistent with the EU Qualified Certificate Statements\")."
+	}
+	return ""
+}
+
+func (l *qcStatemPsd2PolicyMandatory) Execute(c *x509.Certificate) *lint.LintResult {
+	errStr := policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpWeb, "QCP-w", util.IdEtsiQcsQctWeb, "Web", false)
+	util.AppendToStringSemicolonDelim(&errStr, policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpNatural, "QCP-n", util.IdEtsiQcsQctEsign, "eSign", false))
+	util.AppendToStringSemicolonDelim(&errStr, policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpNaturalQscd, "QCP-n-qscd", util.IdEtsiQcsQctEsign, "eSign", true))
+	util.AppendToStringSemicolonDelim(&errStr, policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpLegal, "QCP-l", util.IdEtsiQcsQctEseal, "eSeal", false))
+	util.AppendToStringSemicolonDelim(&errStr, policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpLegalQscd, "QCP-l-qscd", util.IdEtsiQcsQctEseal, "eSeal", true))
+	if errStr != "" {
+		return &lint.LintResult{Status: lint.Error, Details: errStr}
+	}
+	_, psd2Present := util.IsQcStatemPresent(c, &util.IdEtsiPsd2Statem)
+	_, QcTypePresent := util.IsQcStatemPresent(c, &util.IdEtsiQcsQcType)
+	if psd2Present && !QcTypePresent {
+		return &lint.LintResult{Status: lint.Error, Details: "EU Qualified Certificate has PSD2 QcStatement but not the mandatory QcType Statement (ETSI EN 319 412-5: \"When the certificate is issued in accordance with Annex III or Annex IV of Regulation (EU) No 910/2014 [i.8], this statement shall be present\")."}
```

Comment:
I'm having trouble finding supporting text that this is an error. That is, there's nothing preventing the PSD2 QCStatement from being present in a non-qualified certificate. It will not have the legal value for purposes of PSD2, but the QCStatement itself doesn't require that the certificate be qualified (at least, AFAICT from reading TS 119 495).

You've already got the requirement that QcType be present on line 38, so I think this check can just be deleted entirely (lines 60-64). Have I misunderstood a requirement?

Reply:
No. I will delete this check and also chage the name of the lint. Actually the whole lint is not really related to PSD2. I guess this slipped in is due to the very first attempts that targeted at PSD2.

Reply:
Great. Sorry it took me so long to trace through the requirements. The profile for Qualified certificates versus the permissiveness for non-qualified definitely makes this messy to figure out.